bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/cli/CliSpec.java

```diff
@@ -0,0 +1,154 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.runtime.internal.cli;
+
+import io.ballerina.runtime.api.PredefinedTypes;
+import io.ballerina.runtime.api.TypeTags;
+import io.ballerina.runtime.api.creators.ErrorCreator;
+import io.ballerina.runtime.api.creators.ValueCreator;
+import io.ballerina.runtime.api.types.ArrayType;
+import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.types.UnionType;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.values.BArray;
+import io.ballerina.runtime.api.values.BMap;
+import io.ballerina.runtime.api.values.BString;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Class that has the cli argument definitions.
+ *
+ * @since 2.0.0
+ */
+public class CliSpec {
+    private final Option option;
+    private final Operand[] operands;
+    private Object[] mainArgs;
+    private final String[] args;
+
+    public CliSpec(Option option, Operand[] operands, String[] args) {
+        //Todo: refactor code and write tests
+        this.option = option;
+        this.operands = operands;
+        this.args = args;
+    }
+
+    public Object[] getMainArgs() {
+        if (option != null) {
+            BMap<BString, Object> record = option.parseRecord(args);
+            int size = (operands.length + 1) * 2 + 1;
+            mainArgs = new Object[size];
+            processOperands(option.getOperandArgs());
+            mainArgs[size - 2] = record;
+            mainArgs[size - 1] = true;
+        } else {
+            mainArgs = new Object[operands.length * 2 + 1];
+            processOperands(parseOperandArgs());
+        }
+        return mainArgs;
+    }
+
+    private List<String> parseOperandArgs() {
+        List<String> operandArgs = new ArrayList<>();
+        boolean delimeterFound = false;
+        for (String arg :   args) {
+            if (!delimeterFound && CliUtil.isOption(arg)) {
+                if (arg.length() == 2) {
+                    delimeterFound = true;
+                    continue;
+                }
+                throw ErrorCreator.createError(StringUtils.fromString("invalid argument '" + arg));
+            }
+            operandArgs.add(arg);
+        }
+        return operandArgs;
+    }
+
+    private Object[] processOperands(List<String> operandArgs) {
+        int opIndex = 0;
+        int retIndex = 0;
+        Object bValue;
+        int i = 0;
+        while (i < operandArgs.size() && opIndex < operands.length) {
+            Operand curOperand = operands[opIndex++];
+            Type typeOp = curOperand.type;
+            if (typeOp.getTag() == TypeTags.ARRAY_TAG) {
+                BArray bArray = ValueCreator.createArrayValue((ArrayType) typeOp, -1);
+                Type elementType = ((ArrayType) typeOp).getElementType();
+                int elementCount = getElementCount(elementType, operands, opIndex);
```

Comment:
We need to handle args which can make ambiguities like int a [], int b, int c =2,